core/color.js

```diff
@@ -0,0 +1,467 @@
+							}
+							return '#' + color.join( '' );
+						} );
+					} else {
+						return null;
+					}
+				},
+				/**
+				* Normalizes hexadecimal notation so that the color string is always 6 characters long and lowercase.
+				*
+				* @param {String} hexColorCode String containing hex colors.
+				* @returns {String} The style data with hex colors normalized.
+				*/
+				normalizeHex: function( hexColorCode ) {
+					//also style string with hex color
+					var styleHexRegExp = /#(([0-9a-f]{3}){1,2})($|;|\s+)/gi;
+
+					if ( hexColorCode.match( styleHexRegExp ) ) {
+						return hexColorCode.replace( styleHexRegExp, function( match, hexColor, hexColorPart, separator ) {
+							var normalizedHexColor = hexColor.toLowerCase();
+
+							if ( normalizedHexColor.length == 3 ) {
+								var parts = normalizedHexColor.split( '' );
+								normalizedHexColor = [ parts[ 0 ], parts[ 0 ], parts[ 1 ], parts[ 1 ], parts[ 2 ], parts[ 2 ] ].join( '' );
+							}
+
+							return '#' + normalizedHexColor + separator;
+						} );
+					} else {
+						return null;
+					}
+
+				},
+				parseInput: function( colorCode ) {
+					var hexStringFromNamedColor = this._.matchStringToNamedColor( colorCode );
+
+					var hexFromHexString = this._.matchStringToHex( colorCode );
+
+					var hexFromRgbOrHsl = this._.rgbOrHslToHex( colorCode );
+
+					this._.hexColorCode = hexStringFromNamedColor || hexFromHexString || hexFromRgbOrHsl || this._.defaultHexColorCode;
+				},
+				rgbOrHslToHex: function( colorCode ) {
+					var colorFormat = colorCode.trim().slice( 0,3 ).toLowerCase();
+
+					if ( colorFormat !== 'rgb' && colorFormat !== 'hsl' ) {
+						return null;
+					}
+
+					//take values
+					var colorNumberValues = colorCode.match( /\d+\.?\d*/g );
+					if ( !colorNumberValues ) {
+						return null;
+					}
+					colorNumberValues = CKEDITOR.tools.array.map( colorNumberValues, function( number ) {
+						return Number( number );
+					} );
+					//extract alpha
+					var alpha = colorNumberValues.length === 4 ? colorNumberValues.pop() : 1;
+
+					if ( colorFormat === 'hsl' ) {
+						colorNumberValues = this._.hslToRgb( colorNumberValues[0], colorNumberValues[1],colorNumberValues[2] );
+					}
+
+					//blend alpha
+					colorNumberValues = this._.blendAlphaColor( colorNumberValues, alpha );
+
+					return this._.rgbToHex( colorNumberValues );
+				},
+				clampValueInRange: function( value, min, max ) {
+					return Math.min( Math.max( value, min ), max );
+				},
+				normalizePercentValue: function( value ) {
+					if ( value > 1 ) {
+						value =  value / 100;
+					}
+
+					return this._.clampValueInRange( value, 0, 1 );
+				},
+				hslToRgb: function( hue, sat, light ) {
+
+					hue = this._.clampValueInRange( hue, 0, 360 );
+					sat = this._.normalizePercentValue( sat );
+					light = this._.normalizePercentValue( light );
+
+					//based on https://www.w3schools.com/lib/w3color.js
```

Comment:
I would say W3C is not really reliable source. Keep also in mind that borrowing code parts from other projects requires proper license and it's compatibility with CKEditor 4 license so we rarely do this. It's good to base on official specifications like w3.org or whatwg.org if it has necessary information or just common sources like in case of alpha blending algorithm below.

Reply:
I rewrite `HSL` to `RGB` and backward based on the wiki description. Proper links provided in methods.

There are some outstanding specifications (like character case in `hex` color or return value in a case of invalid input string). I assume that this way could be more useful in code than implementing plain specification which may lead to many `undefined` checks.